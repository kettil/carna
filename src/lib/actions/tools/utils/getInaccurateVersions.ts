import { isObject, objectEntries } from '@kettil/tool-lib';
import { coerce } from 'semver';

const regexpPostfix = /\d+\.\d+\.\d+[-.][\d\-.a-z]+/iu;

const getInaccurateVersions = (dependencies: unknown): Array<[string, string, string]> => {
  if (!isObject(dependencies)) {
    return [];
  }

  return objectEntries(dependencies)
    .filter((value): value is readonly [string, string] => typeof value[1] === 'string')
    .filter(([, version]) => !version.startsWith('file'))
    .filter(([, version]) => !version.startsWith('http'))
    .filter(([, version]) => !version.startsWith('git'))
    .map<[string, string, string]>(([name, version]) => [
      name,
      version,
      coerce(version)?.version ?? 'could not be detected',
    ])
    .filter(([, version, result]) => version !== result)
    .filter(([, version]) => !regexpPostfix.test(version));
};

export { getInaccurateVersions };
