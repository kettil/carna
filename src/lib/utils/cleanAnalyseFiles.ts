import { uniqueArray } from '@kettil/tool-lib';
import { eslintExtensions, prettierExtensions } from '../../configs/actionConfigs';

const testEslint = new RegExp(`(${eslintExtensions.replaceAll(',', '|')})$`, 'u');
const testPrettier = new RegExp(`(${prettierExtensions.replaceAll(',', '|')})$`, 'u');

const cleanAnalyseFiles = (
  files: string[],
): { eslintFiles: string[]; prettierFiles: string[]; mergedFiles: string[] } => {
  const eslintFiles = files.filter((file) => testEslint.test(file));
  const prettierFiles = files.filter((file) => testPrettier.test(file));
  const mergedFiles = uniqueArray([...prettierFiles, ...eslintFiles]);

  return { eslintFiles, prettierFiles, mergedFiles };
};

export { cleanAnalyseFiles };
