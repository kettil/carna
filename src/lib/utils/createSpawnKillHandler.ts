import { constants } from 'os';
import { getStdin } from '../cli/process';

type SpawnKillCallback = (signal: NodeJS.Signals | number | undefined) => boolean;

type SpawnKillHandler = {
  addCallback: (callback: SpawnKillCallback) => void;
  kill: (signal: NodeJS.Signals | number | undefined) => boolean[];
};

const createSpawnKillHandler = ({ registerStdin = false }: { registerStdin?: boolean } = {}): SpawnKillHandler => {
  const callbacks: SpawnKillCallback[] = [];
  let isRunning = true;

  if (registerStdin) {
    const stdin = getStdin();

    stdin.setRawMode(true);
    stdin.setEncoding('utf8');
    stdin.resume();

    stdin.on('data', (key) => {
      // ctrl-c ( end of text )
      if (key.toString() === '\u0003') {
        stdin.setRawMode(false);
        stdin.pause();

        callbacks.forEach((callback) => callback(constants.signals.SIGINT));
      }
    });
  }

  return {
    addCallback: (callback) => callbacks.push(callback),
    kill: (signal) => {
      if (isRunning) {
        isRunning = false;

        return callbacks.map((callback) => callback(signal));
      }

      return [];
    },
  };
};

export type { SpawnKillHandler };
export { createSpawnKillHandler };
