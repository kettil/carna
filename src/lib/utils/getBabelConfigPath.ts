import { babelConfigFiles } from '../../configs/actionConfigs';
import { FirstExistFileError } from '../errors/firstExistFileError';
import { getFirstExistingFile } from './getFirstExistingFile';

const getBabelConfigPath = async ([cwd, ...paths]: string[]): Promise<string> => {
  try {
    return await getFirstExistingFile({ cwd, files: babelConfigFiles });
  } catch (error) {
    if (error instanceof FirstExistFileError && paths.length > 0) {
      return getBabelConfigPath(paths);
    }

    throw error;
  }
};

export { getBabelConfigPath };
