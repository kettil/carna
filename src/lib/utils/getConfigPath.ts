import { babelConfigFiles } from '../../configs/actionConfigs';
import { getFirstExistingFile } from './getFirstExistingFile';

const getBabelConfigPath = async ({ cwd }: { cwd: string }): Promise<string> =>
  getFirstExistingFile({ cwd, files: babelConfigFiles });

export { getBabelConfigPath };
