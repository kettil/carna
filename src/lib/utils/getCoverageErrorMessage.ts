import { bold, red } from 'chalk';
import type { CoverageSummary } from 'istanbul-lib-coverage';
import type { Watermark, Watermarks } from 'istanbul-lib-report';
import { getCoverageWatermark } from './getCoverageWatermark';

const getCoverageErrorMessage = (
  key: keyof Watermarks,
  watermarks: Partial<Record<keyof Watermarks, Watermark | number>>,
  coverageSummary: CoverageSummary,
): string => {
  const limit = Math.min(...getCoverageWatermark(watermarks[key]));
  const value = coverageSummary[key].pct;

  return red(`Coverage threshold for ${bold(key)} (${limit}%) not met: ${value}%`);
};

export { getCoverageErrorMessage };
