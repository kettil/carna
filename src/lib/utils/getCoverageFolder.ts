import { join } from 'path';
import { JestActionProps } from '../actions/types';

const getCoverageFolder = (path: string, projects: JestActionProps['projects']): string =>
  join(path, 'coverage', projects.length === 1 ? `_${projects[0].replace(/[^\da-z]+/giu, '_')}` : '');

export { getCoverageFolder };
