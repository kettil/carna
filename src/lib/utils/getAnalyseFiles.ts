import { join } from 'path';
import { promisify } from 'util';
import { isString, uniqueArray } from '@kettil/tool-lib';
import { glob } from 'glob';
import { gitLsAction } from '../actions/git/ls';
import { gitStagedAction } from '../actions/git/staged';
import { access } from '../cmd/access';
import { existFiles } from '../cmd/existFiles';
import type { Task } from '../types';
import { cleanAnalyseFiles } from './cleanAnalyseFiles';

const globPromise = promisify(glob);

type AnalyseFiles = {
  prettierFiles: string[];
  eslintFiles: string[];
};

const getAnalyseFiles: Task<{ path?: string; all?: boolean; files?: AnalyseFiles }, AnalyseFiles | undefined> = async (
  argv,
  { files, all, path },
) => {
  if (all || files) {
    return files;
  }

  if (isString(path)) {
    // get all files from "path"
    const pathFiles = await globPromise(`${path}/**/*`, { cwd: argv.cwd, absolute: true });

    return cleanAnalyseFiles(pathFiles);
  }

  const hasGitFolder = await access(join(argv.root, '.git'), 'readable');

  if (!hasGitFolder) {
    // project has not a git repo => check all files
    return undefined;
  }

  const [changedFiles, stagedFiles] = await Promise.all([
    gitLsAction(argv, { mode: 'all' }),
    gitStagedAction(argv, {}),
  ]);

  const changedAndExistFiles = await existFiles(uniqueArray([...changedFiles, ...stagedFiles]), argv.root);

  return cleanAnalyseFiles(changedAndExistFiles);
};

export type { AnalyseFiles };
export { getAnalyseFiles };
