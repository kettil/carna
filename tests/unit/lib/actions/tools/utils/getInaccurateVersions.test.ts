import { getInaccurateVersions } from '../../../../../../src/lib/actions/tools/utils/getInaccurateVersions';

describe('getInaccurateVersions()', () => {
  test('it should work', () => {
    const result = getInaccurateVersions({
      version: '2.0.1',
      range1: '1.0.0 - 2.9999.9999',
      range2: '>=1.0.2 <2.1.2',
      range3: '>1.0.2 <=2.3.4',
      range4: '<1.0.0 || >=2.3.1 <2.4.5 || >=2.5.2 <3.0.0',
      postfix1: '1.2.3-234',
      postfix2: '1.2.3-alpha',
      postfix3: '1.2.3.beta',
      tilde1: '~1.2',
      tilde2: '~1.2.3',
      wildcard1: '2.x',
      wildcard2: '3.3.x',
      latest: 'latest',
      file: 'file:../common',
      http: 'http://exmplae.com/package.tar.gz',
      git: 'git://github.com/npm/cli.git',
    });

    expect(result).toMatchSnapshot();
  });

  test('it should works with a non object parameter', () => {
    const result = getInaccurateVersions('');

    expect(result).toEqual([]);
  });
});
