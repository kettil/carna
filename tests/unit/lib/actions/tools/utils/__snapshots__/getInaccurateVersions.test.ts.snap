// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInaccurateVersions() it should work 1`] = `
Array [
  Array [
    "range1",
    "1.0.0 - 2.9999.9999",
    "1.0.0",
  ],
  Array [
    "range2",
    ">=1.0.2 <2.1.2",
    "1.0.2",
  ],
  Array [
    "range3",
    ">1.0.2 <=2.3.4",
    "1.0.2",
  ],
  Array [
    "range4",
    "<1.0.0 || >=2.3.1 <2.4.5 || >=2.5.2 <3.0.0",
    "1.0.0",
  ],
  Array [
    "tilde1",
    "~1.2",
    "1.2.0",
  ],
  Array [
    "tilde2",
    "~1.2.3",
    "1.2.3",
  ],
  Array [
    "wildcard1",
    "2.x",
    "2.0.0",
  ],
  Array [
    "wildcard2",
    "3.3.x",
    "3.3.0",
  ],
  Array [
    "latest",
    "latest",
    "could not be detected",
  ],
]
`;
